### PR TITLE
refactor: adapt Portal, Grandpa FV and XDNS to new LightClient trait

### PR DESCRIPTION
**Pull request by MaciejBaj**
# Summary of changes

**Summarize** the description of your changes. 

If it fixes a bug or creates a feature, link the issue.

## Acceptance Checklist

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue):
    - [ ] Have you **written tests **to protect against future occurrences of the bug?
- [ ] New feature (non-breaking change which adds functionality)
    - [ ] Have you **seen it working** in a development environment?
    - [ ] Have you **written tests** for **happy** and **unhappy** paths?
    - [ ] Have you implemented this feature as per the **issue acceptance criteria**?
- [ ] Breaking change (a feature that would cause existing functionality not to work as expected
    - [ ] Is the breaking change **documented**?
    - [ ] Has any previous changes been **deprecated**?
- [ ] CI/CD
    - [ ] If introducing new actions, have you **looked at the action code** for anything spurious?
    - [ ] Have you written **custom scripts** for things that could be actions already on the marketplace?
    - [ ] If introducing new actions, have you **pegged a static version**?
- [ ] Documentation Update
    - [ ] Have you checked other documentation, such as the docs repository?
    - [ ] If updating script documentation, have you **tested it on both environments**?
- [ ] Substrate
    - [ ] RPC methods?
    - [ ] Chainspec, both JSON specs & the module?
    - [ ] Runtime versioning?
    - [ ] Benchmarks?
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

Feel free to explain in further detail your choices if this is a significant change.

### CODE REVIEW

File: primitives/src/light_client.rs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    pub last_updated_height: T::BlockNumber,
+    pub is_halted: bool,
+}
+
+pub trait LightClient<T: frame_system::Config> {
+    fn get_latest_finalized_header(&self) -> Result<Option<Bytes>, DispatchError>;
+
+    fn get_latest_finalized_height(&self) -> Result<Option<T::BlockNumber>, DispatchError>;
+
+    fn get_latest_updated_height(&self) -> Result<Option<T::BlockNumber>, DispatchError>;
```
Comment (AwesomeIbex) [inline primitives/src/light_client.rs:28]: Result<Option> is an anti pattern, i'd suggest doing Result<Enumofyourchoice> for varying states

File: primitives/src/xdns.rs [hunk 2]
```diff
@@ -33,6 +33,46 @@ pub struct FetchXdnsRecordsResponse<AccountId> {
...
+#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
+pub struct TokenRecord {
+    /// Its token 4b Id
+    pub token_id: [u8; 4],
+
+    /// Link to the gateway token is whitelisted on.
+    pub gateway_id: [u8; 4],
+
+    /// Token properties - decimals, symbol, name
+    pub token_props: TokenSysProps,
```
Comment (AwesomeIbex) [inline primitives/src/xdns.rs:47]: How does any of this differ to pallet-assets/registry? Is this not just doing the same thing

Comment (MaciejBaj) [reply primitives/src/xdns.rs:47]: Very possibly 👍

Comment (MaciejBaj) [reply primitives/src/xdns.rs:47]: what would u suggest then, XDNS with the access to pallet-assets or vice versa? 

Circuit should verify if asset is registered at SFX validation before submission. For now Circuit access XDNS

File: types/abi/src/sfx_abi.rs [hunk 3]
```diff
@@ -66,23 +72,26 @@ impl SFXAbi {
         &self,
         ordered_args: &Vec<Data>,
         ordered_args_codec: &Codec,
-    ) -> Result<(), DispatchError> {
+    ) -> Result<FilledAbi, DispatchError> {
         self.ensure_arguments_order(ordered_args)?;
 
+        let prefix_memo: u8 = match self.maybe_prefix_memo {
```
Comment (ahkohd) [inline types/abi/src/sfx_abi.rs]: This line could use a refactor:
```rust
let prefix_memo: u8 = self.maybe_prefix_memo.unwrap_or(0u8);
```

File: types/abi/src/sfx_abi.rs [hunk 4]
```diff
@@ -100,6 +109,22 @@ impl SFXAbi {
         let filled_named_abi: FilledAbi =
             FilledAbi::try_fill_abi(abi, received_payload, payload_codec.clone())?;
 
+        // Check prefix memo if it's set - it's optional since not required for any Event decoding besides Substrate Events
+        // At the same time imposes security risk by attacker faking events sent out of unauthorized pallets
+        if self.maybe_prefix_memo.is_some() {
```
Comment (ahkohd) [inline types/abi/src/sfx_abi.rs]: Umm, this should be collapsed!

File: primitives/src/xdns.rs (hunk 2 above)
Comment (AwesomeIbex) [reply primitives/src/xdns.rs:47]: I would suggest that the circuit should query pallet-assets for existence of our version of that asset, XDNS could also map to the asset index if it **really** wants to, but everything should live in assets/asset-registry

File: primitives/src/xdns.rs [hunk 5]
```diff
@@ -33,6 +33,43 @@ pub struct FetchXdnsRecordsResponse<AccountId> {
...
+}
+
+/// A preliminary representation of a xdns_record in the onchain registry.
+#[derive(Clone, Encode, Decode, Eq, PartialEq, Debug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
+pub struct GatewayRecord<AccountId> {
```
Comment (petscheit) [inline primitives/src/xdns.rs:53]: One thing that is kind of missing here is info on the gateway's native asset, so decimals, and ideally also symbol.

**Pull request merged**